src/app/games/lasca/LascaState.ts

```diff
@@ -0,0 +1,167 @@
+import { Coord } from 'src/app/jscaip/Coord';
+import { GameStateWithTable } from 'src/app/jscaip/GameStateWithTable';
+import { Player } from 'src/app/jscaip/Player';
+import { ArrayUtils, Table } from 'src/app/utils/ArrayUtils';
+import { MGPFallible } from 'src/app/utils/MGPFallible';
+import { Utils } from 'src/app/utils/utils';
+
+export class LascaPiece {
+    public static readonly ZERO: LascaPiece = new LascaPiece(Player.ZERO, false);
+    public static readonly ONE: LascaPiece = new LascaPiece(Player.ONE, false);
+    public static readonly ZERO_OFFICER: LascaPiece = new LascaPiece(Player.ZERO, true);
+    public static readonly ONE_OFFICER: LascaPiece = new LascaPiece(Player.ONE, true);
+
+    public static getPlayerOfficer(player: Player): LascaPiece {
+        if (player === Player.ZERO) {
+            return LascaPiece.ZERO_OFFICER;
+        } else {
+            return LascaPiece.ONE_OFFICER;
+        }
+    }
+    private constructor(public readonly player: Player, public readonly isOfficer: boolean) {}
+
+    public toString(): string {
+        switch (this) {
+            case LascaPiece.ZERO: return 'u';
+            case LascaPiece.ONE: return 'v';
+            case LascaPiece.ZERO_OFFICER: return 'O';
+            default:
+                Utils.expectToBe(this, LascaPiece.ONE_OFFICER);
+                return 'X';
+        }
+    }
+}
+export class LascaSpace {
+
+    public static EMPTY: LascaSpace = new LascaSpace([]);
+
+    public constructor(private readonly pieces: readonly LascaPiece[]) {}
+
+    public isEmpty(): boolean {
+        return this.pieces.length === 0;
+    }
+    public isCommandedBy(player: Player): boolean {
+        if (this.isEmpty()) {
+            return false;
+        }
+        return this.getCommander().player === player;
+    }
+    public getCommander(): LascaPiece {
+        return this.pieces[0];
+    }
+    public getPiecesUnderCommander(): LascaSpace {
+        return new LascaSpace(ArrayUtils.copyImmutableArray(this.pieces).slice(1));
+    }
+    public capturePiece(piece: LascaPiece): LascaSpace {
+        return new LascaSpace(ArrayUtils.copyImmutableArray(this.pieces).concat(piece));
+    }
+    public addStackBelow(stack: LascaSpace): LascaSpace {
+        const piecesBelow: LascaPiece[] = ArrayUtils.copyImmutableArray(stack.pieces);
```

Comment:
pourquoi copier si tu itère juste dessus ?

Reply:
voilà, mgp-set-ifié